src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,523 @@
+                std::float_t tfidf = tf * idf;
+
+                //calculate normVal
+                if(Norm == "l1") {
+                    normVal += abs(tfidf);
+                } else if (Norm == "l2") {
+                    normVal += tfidf * tfidf;
+                }
+
+                //temperarily put output in a vector for future normalization
+                results.emplace_back(std::make_tuple(labelIter->second, tfidf));
+            }
+        }
+        //normVal will never be 0 as long as results is not empty
+        if (normVal == 0.0f)
+            throw std::runtime_error("This happens when the input document is empty");
+
+        // l2-norm calibration
+        if (Norm == "l2")
+            normVal = sqrt(normVal);
+
+        std::vector<SparseVectorEncoding<std::float_t>::ValueEncoding> sparseVector;
+
+        for (auto & result : results) {
+            sparseVector.emplace_back(SparseVectorEncoding<std::float_t>::ValueEncoding(std::get<1>(result) / normVal, std::get<0>(result)));
+        }
+
+        std::sort(sparseVector.begin(), sparseVector.end(),
+            [](SparseVectorEncoding<std::float_t>::ValueEncoding const &a, SparseVectorEncoding<std::float_t>::ValueEncoding const &b) {
+                return a.Index < b.Index;
+            }
+        );
+
+        callback(SparseVectorEncoding<std::float_t>(Labels.size(), std::move(sparseVector)));
+    }
+
+};
+
+namespace Details {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         TfidfVectorizerEstimatorImpl
+///  \brief         Estimator that uses the output of the
+///                 InverseDocumentFrequencyEstimator to provide useful
+///                 information which helps calculation of TfidfVectorizer
+///
+template <size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()>
+class TfidfVectorizerEstimatorImpl : public TransformerEstimator<std::string, SparseVectorEncoding<std::float_t>> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                          = TransformerEstimator<std::string, SparseVectorEncoding<std::float_t>>;
+    using TransformerType                   = TfidfVectorizerTransformer;
+    using IndexMap                          = TfidfVectorizerTransformer::IndexMap;
+    using FrequencyMap                      = TfidfVectorizerTransformer::FrequencyMap;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerEstimatorImpl(
+        AnnotationMapsPtr pAllColumnAnnotations,
+        size_t colIndex,
+        bool binary,
+        std::string norm,
+        bool use_idf,
+        bool smooth_idf,
+        bool sublinear_tf
```

Comment:
These bools would be more easily expressed in terms of enum bit flags rather than individual items.